Feature: URL links on infrastructure page

Kaotic3 opened the pull request:
## Describe your changes

I made it so that you can add an optional URL on the Infrastructure Create or Modify pages, which would then become a clickable link on the status page for that monitor.

## Write your issue number after "Fixes "

#2454 

## Please ensure all items are checked off before requesting a review. "Checked off" means you need to add an "x" character between brackets so they turn into checkmarks.

- [x] (Do not skip this or your PR will be closed) I deployed the application locally.
- [x] (Do not skip this or your PR will be closed) I have performed a self-review and testing of my code.
- [x] I have included the issue # in the PR.
- [ ] I have added i18n support to visible strings (instead of `<div>Add</div>`, use): 
```Javascript
const { t } = useTranslation();
<div>{t('add')}</div>
```
- [x] I have **not** included any files that are not related to my pull request, including package-lock and package-json if dependencies have not changed
- [x] I didn't use any hardcoded values (otherwise it will not scale, and will make it difficult to maintain consistency across the application).
- [x] I made sure font sizes, color choices etc are all referenced from the theme. I don't have any hardcoded dimensions.
- [x] My PR is granular and targeted to one specific feature.
- [x] I ran `npm run format` in server and client directories, which automatically formats your code.
- [x] I took a screenshot or a video and attached to this PR if there is a UI change.

Here are some screen shots of the changes to the UI:

The infrastructure screen has not changed:

![image](https://github.com/user-attachments/assets/4f6db033-44d3-4971-9515-70d6779ce127)

This is what it looks like if you do not add an additional URL:

![image](https://github.com/user-attachments/assets/97cf73c8-576e-4476-a5c2-ff371c3a121d)

This is what it looks like if you edit the monitor that you have:

![image](https://github.com/user-attachments/assets/570f2761-a245-47eb-88ae-b4e9f40ec8c4)

Add a "url" for your application:

![image](https://github.com/user-attachments/assets/7b6dbede-9d12-40b1-b9e9-893e2cf8627f)

How it looks after you add it:

![image](https://github.com/user-attachments/assets/b2f27f26-09f2-472d-bbfa-fc584d5dc525)

This link opens up your application on a new page.

I feel the same could be added to the  Uptime page too, but wanted to try and keep this rather large change as direct as possible.  The monitor.link.url now exists though, so could be added elsewhere?

gorkem-bwl commented:
Quickly glanced on the phone and two things:

- we need to fix the link color.
- we need to identify what a link is. User might be confused why (even optional) they are asked for another url. The explanation can be under the text entry field

Kaotic3 commented:
To hopefully meet the requirements:

"error={errors["linkUrl"] ? true : false}"

This was optional, so I followed the chunk above: "error={errors["name"]  - I thought that was how you were handling optional values, if this is incorrect let me know.

"helperText={errors["linkUrl"]}"

Oh that sent me down a rabbit hole...URL validation wasn't what I thought it was - so I improved it - I moved that long Git regex to the top as an inline const - and I amended it slightly to handle the linkUrl in particular so that it no longer excludes local IPs but otherwise acts the same for a strong validation of the URL.

"we need to fix the link color."

That is just the default, I have not changed it, as I don't know which colour you want - but assuming that would be a simple thing to fix.

"we need to identify what a link is. User might be confused why (even optional) they are asked for another url. The explanation can be under the text entry field"

I added some language to the en.json in locales that goes under the "General Settings":

Here you can select the URL of the host. A friendly name and a URL that will open an application's web page and the authorization secret to connect to the server agent.

The server you are monitoring must be running Capture


You will note that I took out "the" as well, because it either needs to be "the Capture agent" or simply "running Capture" - "running the Capture" that is just not right.

gorkem-bwl commented:
I am good with your proposals here. When it comes to link, it could be the same color as the normal text, but when user hovers on it, the styling can be the same as the one under Settings > About

<img width="1133" alt="image" src="https://github.com/user-attachments/assets/da276c10-f165-4bcc-9c9d-db590ea2a974" />

Can you come up with another screenshot when you fix the link? - to see how it looks + the description.

gorkem-bwl commented:
Rethinking again, why don't we use the same URL and need a different URL to check the service and add another section for this in settings? We can always add the link to the checked URL.

Kaotic3 commented:
I did think myself - why not make the Name the link to the service, and leave the URL for Capture as it is - not a link....I just wanted to keep it neat.

gorkem-bwl commented:
> I did think myself - why not make the Name the link to the service, and leave the URL for Capture as it is - not a link....I just wanted to keep it neat.

Yep, let's remove the extra settings.

Re: image: This is neat, I'll just need to figure out the new underline color and we are done here.

Thanks!

gorkem-bwl commented:
> Do you think replacing the "url" with the "linkUrl" under the name makes sense?
> 
> Don't strip out of the http(s) either and have it as the linkUrl - the api/v1/metrics - doesn't really tell you anything that you didn't already know - as you had to enter it on the page to create the monitor and having the link there with a http references as entered - would make it more obviously a link - and in most cases applications come with a port number and even an /admin /login page (all of which is accepted by the linkUrl).
> 
> I think that might work - and I think leaving it as this colour would be fine then too - as it would match other URL links in the App and would obviously be a link with the http(s) and the application address.


If you mean to include just the URL part (stripping down the /api/... plus the port) would make sense, yes. The link can go to the URL itself and not the port/directory. Am I on the same page with you? :) 

<img width="385" alt="image" src="https://github.com/user-attachments/assets/7dae1ed9-ca57-4199-ac7a-19cde19af676" />

Kaotic3 commented:
Not sure, my view would be that under the Name of the monitor you would have:

http://192.168.1.123:5050/web

or

https://myhosted-item.domain.com/web

Which would then be a clickable URL that took you to the UI for whatever application you were monitoring.  It is optional, if you are monitoring a ping to a machine with 10-hosted items, then add nothing and you get what you get now.

I think the whole http at the start provides enough information for the people using the site to understand that this is a link (and separate it from the standard non http language), there is a fair amount of set-up and knowledge required...that I feel that this assumption isn't expecting too much.

gorkem-bwl commented:
> Not sure, my view would be that under the Name of the monitor you would have:
> 
> http ://192.168.1.123:5050/web
> 
> or
> 
> https ://myhosted-item.domain.com/web
> 
> Which would then be a clickable URL that took you to the UI for whatever application you were monitoring. It is optional, if you are monitoring a ping to a machine with 10-hosted items, then add nothing and you get what you get now.
> 
> I think the whole http at the start provides enough information for the people using the site to understand that this is a link (and separate it from the standard non http language), there is a fair amount of set-up and knowledge required...that I feel that this assumption isn't expecting too much.

OK let me think on this a bit more due to some edge cases. My initial thought is not to introduce another settings for this and be able to directly link (part of the) URL but looks like this is apparently a bit complex due to different setups. 

The screenshot below looks ok to me. For this one, can you please modify your PR so that when you hover on it, there is a tooltip saying something along the lines of "Go to this URL on another tab". Our charts have this tooltip but I doubt we have a relevant component, so you may want to add that simple tooltip component as well. This will bring extra clarity. WDYT?

<img width="502" alt="image" src="https://github.com/user-attachments/assets/8e00bdb2-82ef-4d6a-96a4-9a4a8ebfc2fd" />

Owaiseimdad commented on client/src/Validation/validation.js:
```diff
@@ -374,6 +338,21 @@ const infrastructureMonitorValidation = joi.object({
+		.allow("")
+		.trim()
+		.custom((value, helpers) => {
+			const localUrl =
```
Why is the so complicated? Why not use the same regex used in url?

If we are going to use this regex alot, lets create a common method which validates the urls.

Kaotic3 replied:
It isn't the same Regex.

The "const urlRegex" which I moved to be a constant at the top, actually excludes locally hosted URLs.  The "localRegex" is a version which includes that local range, as you require local IPs.

My initial plan was to make this a single const regex at the top and use the cleaner version through-out, but I realised that it simply didn't work for self-hosted links.

Owaiseimdad reviewed (changes requested):
Can we maintain only one regex for urls? The local-url regex is something I am not able to understand.

Kaotic3 commented:
The end goal is really just to accept any valid URL, including local ranges and any additional pointers - in one of my examples above I provided the link to Pi-Hole, a very common self-hosted application, but to reach it, you must have the IP / Domain - but also /admin/login - without those parts, you receive a 404 error on your link - given our validation step requires a valid URL, we needed a Regex that accepted all of it.

You can of course refactor it to a separate utility and use a flag instead to return one or the other Regex, I didn't take it that far as my view of contributing is always to make minimal changes to ensure that people can visualise and accept those changes without the additional concern of breaking changes due to large scale refactoring - refactoring can always come along latter on - and is going to be required (in my view) because this is an option that many people would probably like on the Uptime / Pagespeed monitors as well as the Infrastructure monitors.

Once the standard is nailed down, then it may be a matter of implementing across all sections and thus doing some major refactoring to make that as simple as possible.  There is a lot of repeated code that I saw that I think could be factored out and replaced by a single line with a flag, which would reduce the overall complexity of individual pages.

@gorkem-bwl that looks a lot more like my original PR, without the "about page" formatting and hover tool-tip.

My only questions are:

1. Why do we want to display the URL for the monitor (/api/v1/metric) is that of any value on this page?
My thinking here is that the Capture is a background task, there is no interacting with it from a user's perspective.
2. As mentioned by @ajhollid - people can put the Capture behind proxies so any value here may bear no relation to the application, it may not have the same domain or ip address.
3. Without the http(s) moniker at the start - it isn't immediately obvious that it is a link and it could easily be missed.

That is my thinking about why I proposed it in the way that I did - that the URL would reflect the link to the application, not to the Capture, and that it would be immediately recognisable as a link with the http(s) prefix.

Let me know which way you decide to go and I will make the necessary changes.
